hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.schema.map.sample;
+
+import com.hazelcast.internal.serialization.Data;
+import com.hazelcast.internal.serialization.InternalSerializationService;
+import com.hazelcast.nio.serialization.ClassDefinition;
+import com.hazelcast.nio.serialization.FieldType;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.nio.serialization.Portable;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.extract.GenericQueryTargetDescriptor;
+import com.hazelcast.sql.impl.extract.QueryPath;
+import com.hazelcast.sql.impl.schema.TableField;
+import com.hazelcast.sql.impl.schema.map.MapTableField;
+import com.hazelcast.sql.impl.type.QueryDataType;
+import com.hazelcast.sql.impl.type.QueryDataTypeUtils;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.LinkedHashMap;
+import java.util.TreeMap;
+
+/**
+ * Helper class that resolves a map-backed table from a key/value sample.
+ */
+public final class MapSampleMetadataResolver {
+
+    private static final String METHOD_PREFIX_GET = "get";
+    private static final String METHOD_PREFIX_IS = "is";
+    private static final String METHOD_GET_FACTORY_ID = "getFactoryId";
+    private static final String METHOD_GET_CLASS_ID = "getClassId";
+
+    private MapSampleMetadataResolver() {
+        // No-op.
+    }
+
+    /**
+     * Resolves the metadata associated with the given key-value sample.
+     *
+     * @param ss Serialization service.
+     * @param target Target to be analyzed.
+     * @param binary Whether map objects are stored in binary form.
+     * @param key Whether passed target is key or value.
+     * @return Sample metadata.
+     * @throws QueryException If metadata cannot be resolved.
+     */
+    public static MapSampleMetadata resolve(
+        InternalSerializationService ss,
+        Object target,
+        boolean binary,
+        boolean key
+    ) {
+        try {
+            if (target instanceof Data) {
+                Data data = (Data) target;
+
+                if (data.isPortable() && binary) {
```

Comment:
Is there any specific reason for using portable style resolution only for binary? If user would change the storage format, we may end up with a different schema.

Reply:
The idea is as follows:
1. Background: in HZ key is always stored as Data, value is stored as either object or key. Portable object exposes different fields depending on whether we process it as object or as Data
1. If the format is OBJECT, we always deserialize to real objects
2. If the format is BINARY, we do not deserialize portables
This works almost in the same way as in Predicate API with one difference: if both key and value as Portables and storage format is OBJECT, then we extract Java fields form both key and value, while in Predicate API we extract portable fields from key, and Java fields from value. 
Our approach looks more consistent to me - user asked us to store key-value pairs In a form of Java objects, and we expose fields of Java objects from both key and value.

Reply:
Another option is to resolve the schema for OBJECT format using the portable definition, but read the values from the Java object itself (I think it was done this way for performance reasons). But that might produce even more weird behavior, looks like we can't do anything about it.

Reply:
The problem is that portable definition may differ from Java fields.